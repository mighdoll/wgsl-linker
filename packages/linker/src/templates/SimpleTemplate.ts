import { Template } from "../ModuleRegistry.js";
import { sliceReplace, sliceWords } from "../Slicer.js";

export const simpleTemplate: Template = {
  name: "simple",
  apply: (src, extParams) => {
    const slices = sliceWords(src, extParams);
    return sliceReplace(src, slices);
  },
};
