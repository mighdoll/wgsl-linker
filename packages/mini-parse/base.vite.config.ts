import { UserConfig } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";
import dts from "vite-plugin-dts";
import { resolve } from "path";

export function baseViteConfig(): UserConfig {
  return {
    plugins: [
      tsconfigPaths(),
      dts(), // generate .d.ts files
    ],
    build: {
      lib: {
        name: "mini-parse",
        entry: [resolve(__dirname, "src/index.ts")],
      },
      minify: false,
      sourcemap: true,
    },
  };
}
