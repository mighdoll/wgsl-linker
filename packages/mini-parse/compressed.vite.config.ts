/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
// import { visualizer } from "rollup-plugin-visualizer";
import { baseViteConfig } from "./vite-base-config.js";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
