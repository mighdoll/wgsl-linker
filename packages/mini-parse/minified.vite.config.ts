/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
// import { visualizer } from "rollup-plugin-visualizer";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
