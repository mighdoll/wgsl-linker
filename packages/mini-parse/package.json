{
  "name": "mini-parse",
  "version": "0.2.0",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "pr": "printenv",
    "build": "run-s build:main build:util build:compressed",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:compressed": "vite -c compressed.vite.config.ts build",
    "build:brotli": "brotli dist/minified.cjs",
    "test": "vitest",
    "organize": "organize-imports-cli tsconfig.json",
    "format": "prettier-eslint \"src/**/*.ts\" --write",
    "lint": "eslint src"
  },
  "type": "module",
  "repository": "github:mighdoll/wgsl-linker",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs",
      "types": "./dist/index.d.ts"
    },
    "./minified": {
      "import": "./dist/minified.js",
      "require": "./dist/minified.cjs",
      "types": "./dist/index.d.ts"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^20.11.16",
    "@typescript-eslint/eslint-plugin": "^6.21.0",
    "@typescript-eslint/parser": "^6.21.0",
    "@vitest/ui": "^1.1.1",
    "berry-pretty": "^0.0.4",
    "brotli": "^1.3.3",
    "esbuild": "^0.20.0",
    "eslint": "^8.56.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "terser": "^5.27.0",
    "typescript": "^5.3.3",
    "vite": "^5.0.12",
    "vite-plugin-dts": "^3.6.4",
    "vite-tsconfig-paths": "^4.2.1",
    "vitest": "^1.2.2"
  }
}
