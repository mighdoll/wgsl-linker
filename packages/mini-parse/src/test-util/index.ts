export * from "./LogCatcher.js";
export * from "./TestParse.js";
