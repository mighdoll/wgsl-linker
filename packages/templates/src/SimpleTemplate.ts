import { Template } from "../../linker/src/ModuleRegistry.js";
import { replaceTokens3 } from "../../linker/src/Util.js";

export const simpleTemplate: Template = {
  name: "simple",
  apply: (src, extParams) => {
    return replaceTokens3(src, extParams);
  }
};
