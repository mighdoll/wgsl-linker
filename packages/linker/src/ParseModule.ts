import { srcLog, SrcMap } from "mini-parse";
import {
  AbstractElem,
  AliasElem,
  ExportElem,
  ExtendsElem,
  FnElem,
  GlobalDirectiveElem,
  ImportElem,
  ModuleElem,
  StructElem,
  TemplateElem,
  VarElem,
} from "./AbstractElems.js";
import { processConditionals } from "./Conditionals.js";
import { ApplyTemplateFn } from "./ModuleRegistry.js";
import { parseWgslD } from "./ParseWgslD.js";
import { SliceReplace, sliceReplace } from "./Slicer.js";

/** module with exportable text fragments that are optionally transformed by a templating engine */
export interface TextModule {
  kind: "text";
  template?: TemplateElem;
  exports: TextExport[];
  fns: FnElem[];
  vars: VarElem[];
  structs: StructElem[];
  imports: (ImportElem | ExtendsElem)[];
  aliases: AliasElem[];
  globalDirectives: GlobalDirectiveElem[];

  /** name of the module. A synthetic name will be assigned if none is provided */
  name: string;

  /** name of the module. A synthetic file name will be assigned if none is provided */
  fileName: string;

  /** original src for module */
  src: string;

  /** src code after processing #if conditionals  */
  preppedSrc: string;

  /** tracks changes through conditional processing for error reporting */
  srcMap: SrcMap;
}

/** an export elem annotated with the fn/struct to which it refers */
export interface TextExport extends ExportElem {
  ref: FnElem | StructElem;
}

let unnamedModuleDex = 0;
let unnamedFileDex = 0;

export function preProcess(
  src: string,
  params: Record<string, any> = {},
  templates: Map<string, ApplyTemplateFn> = new Map()
): SrcMap {
  const condSrcMap = processConditionals(src, params);
  return applyTemplate(condSrcMap, templates, params);
}

export function parseModule(
  src: string,
  templates: Map<string, ApplyTemplateFn> = new Map(),
  fileName = `/unnamed-${unnamedFileDex++}`,
  params: Record<string, any> = {},
  defaultModuleName?: string
): TextModule {
  const srcMap = preProcess(src, params, templates);

  const preppedSrc = srcMap.dest;
  const parsed = parseWgslD(preppedSrc, srcMap);
  const exports = findExports(parsed, srcMap);
  const fns = filterElems<FnElem>(parsed, "fn");
  const aliases = filterElems<AliasElem>(parsed, "alias");
  const globalDirectives = filterElems<GlobalDirectiveElem>(
    parsed,
    "globalDirective"
  );
  const imports = parsed.filter(
    (e) => e.kind === "import" || e.kind === "extends"
  ) as (ImportElem | ExtendsElem)[];
  const structs = filterElems<StructElem>(parsed, "struct");
  const vars = filterElems<VarElem>(parsed, "var");
  const template = filterElems<TemplateElem>(parsed, "template")?.[0];
  matchMergeImports(parsed, srcMap);
  const moduleName = filterElems<ModuleElem>(parsed, "module")[0]?.name;
  matchMergeImports(parsed, srcMap);

  const name = moduleName ?? defaultModuleName ?? `module${unnamedModuleDex++}`;
  const kind = "text";
  return {
    ...{ kind, src, srcMap, preppedSrc, fileName, name },
    ...{
      exports,
      fns,
      structs,
      vars,
      imports,
      template,
      aliases,
      globalDirectives,
    },
  };
}

export function filterElems<T extends AbstractElem>(
  parsed: AbstractElem[],
  kind: T["kind"]
): T[] {
  return parsed.filter((e) => e.kind === kind) as T[];
}

function findExports(parsed: AbstractElem[], srcMap: SrcMap): TextExport[] {
  const results: TextExport[] = [];
  const exports = findKind<ExportElem>(parsed, "export");

  exports.forEach(([elem, i]) => {
    let next: AbstractElem | undefined;
    do {
      next = parsed[++i];
    } while (next?.kind === "extends");
    if (elem.kind === "export") {
      if (next?.kind === "fn" || next?.kind === "struct") {
        results.push({ ...elem, ref: next });
      } else {
        srcLog(srcMap, elem.start, `#export what? (#export a fn or struct)`);
      }
    }
  });
  return results;
}

/** fill in extendsElem field of structs */
function matchMergeImports(parsed: AbstractElem[], srcMap: SrcMap): void {
  const extendsElems = findKind<ExtendsElem>(parsed, "extends");
  extendsElems.forEach(([extendsElem, i]) => {
    let next: AbstractElem | undefined;
    do {
      next = parsed[++i];
    } while (next?.kind === "extends" || next?.kind === "export");
    if (next?.kind === "struct") {
      next.extendsElems = next.extendsElems ?? [];
      next.extendsElems.push(extendsElem);
    } else {
      srcLog(srcMap, extendsElem.start, `#extends not followed by a struct`);
    }
  });
}

function findKind<T extends AbstractElem>(
  parsed: AbstractElem[],
  kind: T["kind"]
): [T, number][] {
  return parsed.flatMap((elem, i) =>
    elem.kind === kind ? ([[elem, i]] as [T, number][]) : []
  );
}

const templateRegex = /#template\s+([/[a-zA-Z_][\w./-]*)/;

function applyTemplate(
  priorSrcMap: SrcMap,
  templates: Map<string, ApplyTemplateFn>,
  params: Record<string, any>
): SrcMap {
  const src = priorSrcMap.dest;
  const foundTemplate = src.match(templateRegex);
  if (!foundTemplate) {
    return priorSrcMap;
  }
  const templateName = foundTemplate[1];
  const templateFn = templates.get(templateName);
  if (!templateFn) {
    srcLog(
      priorSrcMap,
      foundTemplate.index!,
      `template '${templateName}' not found in ModuleRegistry`
    );
    return priorSrcMap;
  }

  // dlog({priorSrcMap})

  const start = foundTemplate.index!;
  const end = start + foundTemplate[0].length;
  const rmDirective: SliceReplace = { start, end, replacement: "" };
  const removedMap = sliceReplace(src, [rmDirective]);
  // dlog({ removedMap });
  const removeMerged = priorSrcMap.merge(removedMap);
  // dlog({ removeMerged });

  const templatedMap = templateFn(removeMerged.dest, params);
  // dlog({ templatedMap});

  const srcMap = removeMerged.merge(templatedMap);
  // dlog({ srcMap });

  return srcMap;
}
