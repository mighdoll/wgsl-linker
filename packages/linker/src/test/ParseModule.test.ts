import { expect, test } from "vitest";
import { parseModule } from "../ParseModule.js";
import { simpleTemplate } from "../templates/SimpleTemplate.js";
import { logCatch } from "mini-parse/test-util";
import { _withBaseLogger } from "mini-parse";

test("simple fn export", () => {
  const src = `
    // #export
    fn one() -> i32 {
      return 1;
    }
  `;
  const module = parseModule(src);
  expect(module.exports.length).toBe(1);
  expect(module).toMatchSnapshot();
});

test("simple fn import", () => {
  const src = `
    // #import foo
    fn bar() { foo(); }
  `;
  const module = parseModule(src);
  expect(module.imports.length).toBe(1);
  expect(module.imports[0].name).toBe("foo");
  expect(module).toMatchSnapshot();
});

test("match #extends", () => {
  const src = `
    // #extends Foo
    // #extends Bar
    struct Elem {
      sum: f32
    }
  `;
  const module = parseModule(src);
  const merges = module.structs[0].extendsElems!;
  expect(merges[0].name).eq("Foo");
  expect(merges[1].name).eq("Bar");
});

test("read simple struct export", () => {
  const exportPrefix = `// #export`;
  const src = `
    struct Elem {
      sum: f32
    }
  `;
  const module = parseModule(exportPrefix + "\n" + src);
  expect(module.exports.length).toBe(1);
  const firstExport = module.exports[0];
  expect(firstExport.ref.name).toBe("Elem");
});

test("read #module", () => {
  const src = `
    // #module my.module.com
    // #export
    fn foo() {}
  `;
  const textModule = parseModule(src);
  expect(textModule.name).toBe("my.module.com");
});

test("simple #template preserves src map", () => {
  const src = `
    // #template simple
    fn foo() { XX }
  `;
  const expected = `
    // 
    fn foo() { /**/ }
  `;
  const templates = new Map([["simple", simpleTemplate.apply]]);
  const textModule = parseModule(src, templates, "./foo", { XX: "/**/" });
  expect(textModule.preppedSrc).includes("fn foo() { /**/ }");
  expect(textModule.preppedSrc).equals(expected);
  expect(textModule.srcMap.entries).length(3);
});

test("parse error shows correct line after simple #template", () => {
  const src = `
    // #template simple
    fn foo () { XX }
    fn () { } // oops
  `;
  const templates = new Map([["simple", simpleTemplate.apply]]);
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => {
    parseModule(src, templates, "./foo", { XX: "/**/" });
  });
  expect(logged()).toMatchInlineSnapshot(`
    "missing fn name
        fn () { } // oops   Ln 4
          ^"
  `);
});

test("parse error shows correct line after #ifdef ", () => {
  const src = `
    // #if FALSE
    foo
    bar
    // #endif
    fn () { } // oops
  `;
  const templates = new Map([["simple", simpleTemplate.apply]]);
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => {
    parseModule(src, templates, "./foo", { XX: "/**/" });
  });
  expect(logged()).toMatchInlineSnapshot(`
    "missing fn name
        fn () { } // oops   Ln 6
          ^"
  `);
});

test("parse error shows correct line after #ifdef and simple #template", () => {
  const src = `
    // #if FALSE
    foo
    bar
    // #endif
    // #template simple
    fn foo () { XX }
    fn () { } // oops
  `;
  const templates = new Map([["simple", simpleTemplate.apply]]);
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => {
    parseModule(src, templates, "./foo", { XX: "/**/" });
  });
  expect(logged()).toMatchInlineSnapshot(`
    "missing fn name
        fn () { } // oops   Ln 8
          ^"
  `);
});
