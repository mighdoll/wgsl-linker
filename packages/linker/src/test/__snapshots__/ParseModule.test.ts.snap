// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`import rust style 1`] = `
[
  {
    "end": 19,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "my",
        },
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "foo",
        },
      ],
    },
    "kind": "treeImport",
    "start": 5,
  },
]
`;

exports[`simple fn export 1`] = `
{
  "aliases": [],
  "exports": [
    {
      "end": 16,
      "kind": "export",
      "ref": {
        "calls": [],
        "end": 59,
        "kind": "fn",
        "name": "one",
        "nameElem": {
          "end": 26,
          "kind": "fnName",
          "name": "one",
          "start": 23,
        },
        "start": 20,
        "typeRefs": [
          {
            "end": 35,
            "kind": "typeRef",
            "name": "i32",
            "start": 32,
          },
        ],
      },
      "start": 8,
    },
  ],
  "fns": [
    {
      "calls": [],
      "end": 59,
      "kind": "fn",
      "name": "one",
      "nameElem": {
        "end": 26,
        "kind": "fnName",
        "name": "one",
        "start": 23,
      },
      "start": 20,
      "typeRefs": [
        {
          "end": 35,
          "kind": "typeRef",
          "name": "i32",
          "start": 32,
        },
      ],
    },
  ],
  "globalDirectives": [],
  "imports": [],
  "kind": "text",
  "modulePath": "./test.wgsl",
  "name": "oldName",
  "preppedSrc": "
    // #export
    fn one() -> i32 {
      return 1;
    }
  ",
  "src": "
    // #export
    fn one() -> i32 {
      return 1;
    }
  ",
  "srcMap": SrcMap {
    "dest": "
    // #export
    fn one() -> i32 {
      return 1;
    }
  ",
    "entries": [
      {
        "destEnd": 62,
        "destStart": 0,
        "src": "
    // #export
    fn one() -> i32 {
      return 1;
    }
  ",
        "srcEnd": 62,
        "srcStart": 0,
      },
    ],
  },
  "structs": [],
  "template": undefined,
  "vars": [],
}
`;

exports[`simple fn import 1`] = `
{
  "aliases": [],
  "exports": [],
  "fns": [
    {
      "calls": [
        {
          "end": 38,
          "kind": "call",
          "name": "foo",
          "start": 35,
        },
      ],
      "end": 43,
      "kind": "fn",
      "name": "bar",
      "nameElem": {
        "end": 30,
        "kind": "fnName",
        "name": "bar",
        "start": 27,
      },
      "start": 24,
      "typeRefs": [],
    },
  ],
  "globalDirectives": [],
  "imports": [
    {
      "end": 19,
      "imports": ImportTree {
        "segments": [
          SimpleSegment {
            "args": undefined,
            "as": undefined,
            "name": "foo",
          },
        ],
      },
      "kind": "treeImport",
      "start": 8,
    },
  ],
  "kind": "text",
  "modulePath": "./test.wgsl",
  "name": "oldName",
  "preppedSrc": "
    // #import foo
    fn bar() { foo(); }
  ",
  "src": "
    // #import foo
    fn bar() { foo(); }
  ",
  "srcMap": SrcMap {
    "dest": "
    // #import foo
    fn bar() { foo(); }
  ",
    "entries": [
      {
        "destEnd": 46,
        "destStart": 0,
        "src": "
    // #import foo
    fn bar() { foo(); }
  ",
        "srcEnd": 46,
        "srcStart": 0,
      },
    ],
  },
  "structs": [],
  "template": undefined,
  "vars": [],
}
`;
